### PR TITLE
[CLI-534] Checks to see if an artifact about to be logged is of the correct type

### PR DESCRIPTION
**Pull request by tssweeney**
This PR reduces chances of user error by checking the artifact type before performing the log operation. Without this change, the log operation fails silently in the background and the user is unaware of the failed log.

Screen Shot 2020-11-25 at 11 11 48 AM

Given the above example, a value error will be thrown on line 9.

Note: it may be up for debate if we want to throw a runtime error. I lean on the side of notifying the user so that they can fix their script, but open to discussion here.

### CODE REVIEW

**Review by vanpelt — APPROVED** [review]

File: standalone_tests/artifact_tests.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    with wandb.init() as run:
+        artifact = run.use_artifact(artifact_name + ":latest")
+        # should work
+        image = artifact.get("image")
+        assert image is not None
+        # should not work
+        image_2 = artifact.get("image_2")
+        assert image_2 is None
+
+    shutil.rmtree("wandb")
```
Comment (shawnlewis) [inline standalone_tests/artifact_tests.py:40]: It's better to put stuff like this in a test teardown function (automatically executed by the unit test framework), to a guarantee cleanup happens, even if the test itself crashes.

File: wandb/sdk/wandb_run.py [hunk 2]
```diff
@@ -1669,9 +1669,28 @@ def log_artifact(self, artifact_or_path, name=None, type=None, aliases=None):
         if isinstance(aliases, str):
             aliases = [aliases]
         artifact.finalize()
+        self._assert_can_log_artifact(artifact)
```
Comment (shawnlewis) [inline wandb/sdk/wandb_run.py:1672]: Does this break offline mode? It should be possible to run artifact logging runs without being connected to the internet: https://docs.wandb.com/library/technical-faq#can-i-run-wandb-offline

**Review by shawnlewis — CHANGES_REQUESTED** [review]
This is a good

File: wandb/sdk/wandb_run.py [hunk 3]
```diff
@@ -1669,9 +1669,28 @@ def log_artifact(self, artifact_or_path, name=None, type=None, aliases=None):
...
         self._backend.interface.publish_artifact(self, artifact, aliases)
         return artifact
 
+    def _assert_can_log_artifact(self, artifact):
+        overrides = {"run": self.id}
+        run_obj = self._run_obj
```
Comment (annirudh) [inline wandb/sdk/wandb_run.py:1678]: We have this logic duplicated now between `use_artifact` and here. Might be worth adding a `_public_api()` factory that handles this.

File: wandb/sdk/wandb_run.py [hunk 4]
```diff
@@ -1669,9 +1669,28 @@ def log_artifact(self, artifact_or_path, name=None, type=None, aliases=None):
         if isinstance(aliases, str):
             aliases = [aliases]
         artifact.finalize()
+        self._assert_can_log_artifact(artifact)
         self._backend.interface.publish_artifact(self, artifact, aliases)
         return artifact
 
+    def _assert_can_log_artifact(self, artifact):
```
Comment (annirudh) [inline wandb/sdk/wandb_run.py:1676]: We need to conditionally call the public API based on whether or not we're in offline mode (which you can get from `_settings._mode` I believe).

If the user is in dryrun / offline mode, we can just skip validation. We have no choice but to error at sync time in that scenario.

File: standalone_tests/artifact_tests.py (hunk 1 above)
Comment (tssweeney) [reply standalone_tests/artifact_tests.py:40]: Great - I pulled into it's own function and ensured it gets called with a try block in the main.

File: wandb/sdk/wandb_run.py (hunk 4 above)
Comment (tssweeney) [reply wandb/sdk/wandb_run.py:1676]: Great, I added this condition in next commit - thanks for pointing this out

File: wandb/sdk/wandb_run.py (hunk 2 above)
Comment (tssweeney) [reply wandb/sdk/wandb_run.py:1672]: Per Anni's comment below, I added a new condition to check for this in offline setting in the next commit

File: wandb/sdk/wandb_run.py (hunk 3 above)
Comment (tssweeney) [reply wandb/sdk/wandb_run.py:1678]: nice call - added in next commit

File: wandb/apis/public.py [hunk 5]
```diff
@@ -2531,6 +2531,46 @@ def aliases(self, aliases):
                 )
         self._aliases = aliases
 
+    @staticmethod
+    def expected_type(client, name, entity_name, project_name):
```
Comment (annirudh) [inline wandb/apis/public.py:2535]: Do we need this? I thought this would be equivalent to:

```
public_api.artifact(name).type
```

**Review by annirudh — APPROVED** [review]
Looks good!

Would you mind that `log_artifact` works in offline mode even if the type is mismatched? And that we get an error when we try to sync it?

File: wandb/apis/public.py (hunk 5 above)
Comment (annirudh) [reply wandb/apis/public.py:2535]: Oh, I guess the difference is in how it handles the case where the artifact doesn't exist, which makes sense.

File: wandb/apis/public.py [hunk 6]
```diff
@@ -2531,6 +2531,46 @@ def aliases(self, aliases):
...
+            },
+        )
+
+        project = response.get("project")
+        if project is not None:
+            artifact = project.get("artifact")
+            if artifact is not None:
+                artifact_type = artifact.get("artifactType")
+                if artifact_type is not None:
+                    return artifact_type.get("name")
```
Comment (annirudh) [inline wandb/apis/public.py:2572]: Should we explicitly return `None` if any of the above are missing?

File: wandb/apis/public.py (hunk 5 above)
Comment (tssweeney) [reply wandb/apis/public.py:2535]: Yeah, which could be handled in a try/except.

File: wandb/apis/public.py (hunk 6 above)
Comment (tssweeney) [reply wandb/apis/public.py:2572]: Sure - I can make it explicit. Incoming.

**Review by shawnlewis — APPROVED** [review]

**Pull request merged**